For #1268 Add Unit test for PrivateShortcutCreateManager

RotBolt opened the pull request:
### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [X] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [X] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

Closes #1268
PR just contains a unit test.
It does not contain any UI related change or accessibility change

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

RotBolt commented:
Hello @Mugurell 
I have made this PR for `PrivateCreateShortcutManager` following your suggestions. 
I have ran `detekt` and `ktlint` as well and they went fine
Added the `pre-push hook` as well.

Could you please review this ?

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+    @Test
+    fun `GIVEN shortcut pinning is not supported WHEN createPrivateShortcut is called THEN do not create a pinned shortcut`() {
+        val privateShortcutCreateManager = spyk<PrivateShortcutCreateManager>()
+        mockkStatic(ShortcutManagerCompat::class)
```
As a general rule, to be sure all static mocks will be unmocked at the end of a test no matter what we can use
-  `mockkStatic` in `@Before` and `unmockkStatic` in `@After` or 
- in a try-catch-finally block like [here](https://github.com/mozilla-mobile/fenix/blob/bbea03e27cb37e0227f434450b77a1bb0be76818/app/src/test/java/org/mozilla/fenix/tabstray/DefaultTabsTrayControllerTest.kt#L260-L278).

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+        val intentSender = slot<IntentSender>()
+        val intent = slot<Intent>()
+
+        val privateShortcutCreateManager = spyk<PrivateShortcutCreateManager>()
```
We probably don't need a spy or even an instance here.

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+        val intent = slot<Intent>()
+
+        val privateShortcutCreateManager = spyk<PrivateShortcutCreateManager>()
+        mockkStatic(ShortcutManagerCompat::class)
```
This static mocks can be put in a try-catch-finally block.

RotBolt replied:
went with `@Before and @After approach`

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+
+        verify { PendingIntent.getActivity(testContext, 0, capture(intent), PendingIntent.FLAG_UPDATE_CURRENT) }
+        verify { ShortcutManagerCompat.requestPinShortcut(testContext, capture(shortcut), capture(intentSender)) }
+        `assert shortcutInfoCompat is build correctly`(shortcut, testContext)
```
- Instead of sending the CapturingSlot maybe it would be cleaner to send it's value?
- If the asserts will use the same static reference to a context they don't really need it to be passed as an argument.

RotBolt replied:
regarding `testContext` , I just did it as function preference to have arguments. But I get it. it doesn't make sense in unit test

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+        verify { PendingIntent.getActivity(testContext, 0, capture(intent), PendingIntent.FLAG_UPDATE_CURRENT) }
+        verify { ShortcutManagerCompat.requestPinShortcut(testContext, capture(shortcut), capture(intentSender)) }
+        `assert shortcutInfoCompat is build correctly`(shortcut, testContext)
+        `assert homeScreenIntent is built correctly`(intent)
```
Maybe it would be better to send the value of the `CapturingSlot` instead?

RotBolt replied:
sent the captured value

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    private fun `assert shortcutInfoCompat is build correctly`(shortcutInfoCompat: CapturingSlot<ShortcutInfoCompat>, context: Context) {
+        shortcutInfoCompat.apply {
```
`apply` is not used in an idiomatic way here.
If you'll send the captured value as a parameter there will not be a need for `shortcutInfoCompat.captured.*` so the code will already be cleaner and not need these scoping functions but if you want to use them maybe [with](https://kotlinlang.org/docs/scope-functions.html#with) is a more appropriate one for a general `_with_ this object, do the following...` scenario.

RotBolt replied:
After sending `captured value` it doesn't make sense to use `with or apply`
updated the code

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    private fun `assert homeScreenIntent is built correctly`(intent: CapturingSlot<Intent>) {
+        intent.apply {
```
The same as above. Maybe `with` is a more appropriate scoping function to use here or maybe we don't even need one.

RotBolt replied:
Yes, after sending captured value directly. It does not make sense 
I have updated the code

Mugurell reviewed (commented):
Thank you @RotBolt !
This looks clean and covers all the functionality we needed to be tested.
Still some minor improvements to be made but it's looking great!

RotBolt commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+package org.mozilla.fenix.components
```
Sure will add

RotBolt commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,83 @@
+
+    @Test
+    fun `GIVEN shortcut pinning is not supported WHEN createPrivateShortcut is called THEN do not create a pinned shortcut`() {
+        val privateShortcutCreateManager = spyk<PrivateShortcutCreateManager>()
```
Yes, I missed that thanks for pointing it out

RotBolt commented:
Hello @Mugurell  , I have addressed your comments.

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,92 @@
+package org.mozilla.fenix.components
+
+import android.app.PendingIntent
+import android.content.Context
```
ktlint: Unused import

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,92 @@
+import android.content.IntentSender
+import androidx.core.content.pm.ShortcutInfoCompat
+import androidx.core.content.pm.ShortcutManagerCompat
+import io.mockk.CapturingSlot
```
ktlint: Unused import

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,92 @@
+import io.mockk.every
+import io.mockk.mockkStatic
+import io.mockk.slot
+import io.mockk.spyk
```
ktlint: Unused import

RotBolt replied:
aah I missed to run . Thanks

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    @After
+    fun tearDown() {
```
Can you please move this above the other public `@Test` methods?.
Seeing `@Before` and `@After` above all tests helps with readability & file navigation.

RotBolt replied:
understood. I preferred at bottom in my personal projects. Thats why added in bottom

Mugurell commented on app/src/test/java/org/mozilla/fenix/components/PrivateShortcutCreateManagerTest.kt:
```diff
@@ -0,0 +1,92 @@
+
+    @Test
+    fun `GIVEN shortcut pinning is not supported WHEN createPrivateShortcut is called THEN do not create a pinned shortcut`() {
+
```
nit: extraneous newline

RotBolt replied:
resolved

Mugurell commented:
> Hello @Mugurell , I have addressed your comments.

Just a bit of cleanup and then we can land this.

RotBolt commented:
> > Hello @Mugurell , I have addressed your comments.
> 
> Just a bit of cleanup and then we can land this.

I ran the ktlint and detekt this time. I think this time all issues are addressed :sweat_smile:

Mugurell reviewed (approved):
Looks great!
Thank you @RotBolt!
🚀

jonalmeida commented:
Removing the 'needs landing' label for now since Mergify isn't proceeding forward for some reason, and a PR queue is blocked as a result.

RotBolt commented:
> Removing the 'needs landing' label for now since Mergify isn't proceeding forward for some reason, and a PR queue is blocked as a result.

Is this issue in the CI pipeline only or is there any step I need to take on my end ?

Mugurell commented:
> > Removing the 'needs landing' label for now since Mergify isn't proceeding forward for some reason, and a PR queue is blocked as a result.
> 
> Is this issue in the CI pipeline only or is there any step I need to take on my end ?

Nothing left to do on your end, only need to merge the changes.
Rebased your PR over the latest master and will try again.
